x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -45,18 +45,37 @@
     private static final Logger logger = LogManager.getLogger(SetSingleNodeAllocateStep.class);
     public static final String NAME = "set-single-node-allocation";
 
-    public SetSingleNodeAllocateStep(StepKey key, StepKey nextStepKey, Client client) {
+    private Integer numberOfShards;
+
+    public SetSingleNodeAllocateStep(StepKey key, StepKey nextStepKey, Client client, Integer numberOfShards) {
         super(key, nextStepKey, client);
+        this.numberOfShards = numberOfShards;
     }
 
     @Override
     public boolean isRetryable() {
         return true;
     }
 
+    public Integer getNumberOfShards() {
+        return numberOfShards;
+    }
+
     @Override
     public void performAction(IndexMetadata indexMetadata, ClusterState clusterState,
                               ClusterStateObserver observer, ActionListener<Boolean> listener) {
+        String indexName = indexMetadata.getIndex().getName();
```

Comment:
What do you think about refactoring all of this into some new `Step`? I like the logic overall, and I agree that we should do the validation before the single allocation starts, I'm just not sure it belongs **within** this step.

Reply:
Yeah, it's a good idea, I will refactor the code soon.

Reply:
hi @joegallo , I've refactor the code yet, can you take a look at the new commit?